### ISSUE TITLE
Add interface to input data layers (lmdb/leveldb/hdf5 etc.) to query dataset size

### ISSUE THREAD

**Issue by shaibagon** [BVLC/caffe#3436]
During training, one usually wants to know the training/validation dataset size.
This information is useful for determining the number of iterations required to "cover" the entire set (aka an "epoch"), to determine the number of test iterations require to cover the validation set, etc.

The way implementation currently stand, user must manually hard code these numbers into his solver prototxt and it is very unclear from looking at a prototxt how many epochs where done to train the net.

It would be nice if [`DataLayer`](https://github.com/BVLC/caffe/blob/master/include/caffe/layers/data_layer.hpp) class would have an extra method like 
`virtual inline int GetNumSamples() const { return -1; }`

It would be even nicer if this method would be accessible through the python/matlab interface.

See, for example, this SO thread: http://stackoverflow.com/q/34092606/1714410

Thank you very much,
Shai

**Comment by Noiredd** [comment]
I don't agree that the `DataLayer` should do that. Any kind of control over the training (i.e. how many epochs to make) should be known *before* the layer is even instantiated, not during training. So if we should do this kind of dataset analysis, it should be the solver who performs it. And this raises several problems - what if we read from multiple databases? How to handle multiple data layers?  
Personally, I'm okay with the way we're currently doing this.
